### ISSUE TITLE
'Null object cannot be converted to a value type.' when using nullable type on interface

### ISSUE THREAD

**Issue by JaseRandall** [StackExchange/Dapper#933]
Using Dapper.Contrib (for the change tracking) and SQLite.

I have an interface and class... here's a cut down version:
```
public interface IExample
{
   public int Id { get; set; }
   DateTime? Dob { get; set; }
}

public class Example : IExample
{
   public int Id { get; set; }
   public DateTime? Dob { get; set; }
}
```

**This works:**
```
public IEnumerable<IExample> GetAll()
{
    return Connection.GetAll<Example>();
}
```

**This does not work:**
```
public IEnumerable<IExample> GetAll()
{
    return Connection.GetAll<IExample>();
}
```

It throws 'Null object cannot be converted to a value type.' which obviously isn't right since I'm using a nullable type.  I've removed all nulls from the database and it made no difference.  If I change DateTime? to DateTime then it works.  I could work around it by assinging DateTime.MinValue however I'd prefer not to.

Is this a bug in Dapper.Contrib or something I'm doing wrong?

**Comment by NickCraver** [comment]
When you're trying to deserialize into `IExample`, what is the type of the object in the list? When you call `.GetType()` on an object, what would you expect the result to be?

I ask the above to illustrate why deserializing to an interface (or an abstract class) wouldn't really work...because you can't instantiate it. Does that make sense?

**Comment by creative999** [comment]
I've had more time today to look into this and have worked out why it occurs when I use the interface.

What you've said does make sense however the issue isn't the interface.  Dapper.Contrib has code to create an object from an interface:

```
if (!type.IsInterface()) return connection.Query<T>(sql, null, transaction, commandTimeout: commandTimeout);

var result = connection.Query(sql);
var list = new List<T>();
 foreach (IDictionary<string, object> res in result)
 {
    var obj = ProxyGenerator.GetInterfaceProxy<T>();
    foreach (var property in TypePropertiesCache(type))
    {
        var val = res[property.Name];
        property.SetValue(obj, Convert.ChangeType(val, property.PropertyType), null);
     }
     ((IProxy)obj).IsDirty = false;   //reset change tracking and return
     list.Add(obj);
}
```
I've never used reflection, am pretty junior in terms of programming experience and I don't program professional however, I think the issue is caused by:
` property.SetValue(obj, Convert.ChangeType(val, property.PropertyType), null);`
which can't set the value of a nullable type.

This link explains why: [https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/nullable-types/how-to-identify-a-nullable-type](url)

I've also looked at [http://codewut.de/content/how-set-nullable-types-reflection-properties-c-net]() which suggests a solution.  I adapted that for Dapper.Contrib and the line above that failed could be changed to:
```
if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>)) 
{
     var genericType = property.PropertyType.GetGenericArguments()[0];
     property.SetValue(obj, Convert.ChangeType(val, genericType), null);
}
else
{
     property.SetValue(obj, Convert.ChangeType(val, property.PropertyType), null);
}

```
I haven't been able to test this so I'm not sure if it's the correct solution.
(I can't figure out how to do the NuGet package restore after downloading the Dapper source code.)

**Comment by JaseRandall** [comment]
Just further to the last:
I've managed to download the Dapper source code and test the solution.
I found a couple of issue... firstly I had DOB in the database rather than Dob.
I fixed that then had this error:
> System.InvalidCastException: 'Invalid cast from 'System.String' to 'System.Nullable`1[[System.DateTime, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]'.'
That's understandable; The date value is stored as a string in SQLite and property.SetValue() can't cast that to nullable.
I changed:
```
var val = res[property.Name];
property.SetValue(obj, Convert.ChangeType(val, property.PropertyType), null);
```
To:
```
var val = res[property.Name];
if (property.PropertyType.IsGenericType() && property.PropertyType.GetGenericTypeDefinition() == typeof(Nullable<>))
{
    if (val != null)
    {
        var genericType = property.PropertyType.GetGenericArguments()[0];
        property.SetValue(obj, Convert.ChangeType(val, genericType), null);
    }
    }
    else
    {
    property.SetValue(obj, Convert.ChangeType(val, property.PropertyType), null);
    }
}
```
and after that change, Dapper.Contrib was able to handle the DateTime? type including whether the value from the database was null or an actual value.

Now I just have to figure out how to do a Pull Request.

**Comment by JaseRandall** [comment]
Issue resolved by Pull Request #936.
Closed.
